rclpy/rclpy/executors.py

```diff
@@ -366,11 +367,35 @@ def spin_once_until_future_complete(
     def _take_timer(self, tmr):
         try:
             with tmr.handle:
-                tmr.handle.call_timer()
+                info = tmr.handle.call_timer_with_info()
+                timer_info = TimerInfo(
+                    expected_call_time=info['expected_call_time'],
+                    actual_call_time=info['actual_call_time'],
+                    clock_type=tmr.clock.clock_type)
+
+                # This needs to be done somewhere else since this is certain overhead to call timer
+                # for every iteration. at this moment stays here for review purpose.
+                # Besides, probably we can make this generic utility function
+                def check_argument_type(callback_func, target_type):
+                    sig = inspect.signature(callback_func)
+                    for param in sig.parameters.values():
+                        print(param)
+                        if param.annotation == target_type:
+                            # return 1st one immediately
+                            return param.name
+                    # We could not find the target type in the signature
+                    return None
 
-                async def _execute():
-                    await await_or_execute(tmr.callback)
-                return _execute
+                arg_name = check_argument_type(tmr.callback, target_type=TimerInfo)
+                prefilled_arg = {arg_name: timer_info}
+                if arg_name is not None:
```

Comment:
How about move this line under if
```suggestion
                if arg_name is not None:
                    prefilled_arg = {arg_name: timer_info}
```

Reply:
@Barry-Xu-2018 thanks for the comment. this can be improved more, i would like to wait for the feedback for this approach before adding the test cases and cosmetic changes. but thanks!